### PR TITLE
Pixel shuffle Layer

### PR DESCRIPTION
**Pull request by abh2k**
Added  pixel shuffle layer

### CODE REVIEW

**Comment by abh2k** [comment]
@zoq @kartikdutt18   Are  there any  more changes required     in this pr.

File: src/mlpack/methods/ann/layer/pixel_shuffle_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+          size_t height_index = h / upscaleFactor;
+          size_t width_index = w / upscaleFactor;
+          size_t channel_index = (upscaleFactor * (h % upscaleFactor)) +
+              (w % upscaleFactor) + (c * std::pow(upscaleFactor, 2));
+          outputTemp(w, h, c) = inputTemp(width_index, height_index,
+              channel_index);
+        }
+      }
+    }
+    output.col(n) = outputImage;
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/pixel_shuffle_impl.hpp:93]:
```suggestion
    }

    output.col(n) = outputImage;
```

File: src/mlpack/methods/ann/layer/pixel_shuffle_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  }
+  output.zeros(outputHeight * outputWidth * sizeOut, batchSize);
+  for (size_t n = 0; n < batchSize; n++)
+  {
+    arma::mat inputImage = input.col(n);
+    arma::mat outputImage = output.col(n);
+    arma::cube inputTemp(const_cast<arma::mat&>(inputImage).memptr(), height,
+        width, size, false, false);
+    arma::cube outputTemp(const_cast<arma::mat&>(outputImage).memptr(),
+        outputHeight, outputWidth, sizeOut, false, false);
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/pixel_shuffle_impl.hpp:76]: I'm not sure about this but maybe we can do this without the creating copies (which add to latency / space utilization). If 
```
arma::cube inputTemp(const_cast<arma::mat&>(input.col(n)).memptr(), height,
          width, size, false, false);
```
doesn't work, we can 
```
arma::cube inputTemp(const_cast<arma::mat&>(input).memptr(), height,
          width, size * batchSize, false, false);
```
and then use span. That way we don't create copies.
Also save time on copying back (maybe). Let me know incase I missed something.

File: src/mlpack/methods/ann/layer/pixel_shuffle_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+          size_t height_index = h / upscaleFactor;
+          size_t width_index = w / upscaleFactor;
+          size_t channel_index = (upscaleFactor * (h % upscaleFactor)) +
+              (w % upscaleFactor) + (c * std::pow(upscaleFactor, 2));
+          gTemp(width_index, height_index, channel_index) = gyTemp(w, h, c);
+        }
+      }
+    }
+    g.col(n) = gImage;
+  }
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/pixel_shuffle_impl.hpp:127]:
```suggestion
    }

    g.col(n) = gImage;
  }
```

**Review by kartikdutt18 — COMMENTED** [review]
Hey @abh2k, These are minor changes and a point to discuss on. Thanks for taking the time to create the PR.
Regards.

File: src/mlpack/methods/ann/layer/pixel_shuffle_impl.hpp (hunk 2 above)
Comment (abh2k) [reply src/mlpack/methods/ann/layer/pixel_shuffle_impl.hpp]: @kartikdutt18 @zoq I made the required changes.  Now the copying step has been removed. Please  review.

**Review by kartikdutt18 — APPROVED** [review]
Looks good to me!
Thanks a lot for the contribution.
Regards.

File: src/mlpack/methods/ann/layer/pixel_shuffle.hpp [hunk 4]
```diff
@@ -0,0 +1,180 @@
+/**
+ * @file methods/ann/layer/pixel_shuffle.hpp
+ * @author Anjishnu Mukherjee
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/pixel_shuffle.hpp:3]: Feel free to add yourself as co-author.

**Review by birm — APPROVED** [review]
Thanks for following up on #2563 :)

**Pull request merged**